File: leetcode/remove-duplicates.go
package main

// https://leetcode.com/problems/remove-duplicates-from-sorted-array/description/
func removeDuplicates(nums []int) int {
	length := len(nums)
	if length < 2 {
		return length
	}
	//maintain two pointers. i for traversal and k for the shift
	k, i := 1, 1
	for i < length {
		//if current and previous are not same then shift element at i to k
		if nums[i] != nums[i-1] {
			nums[k] = nums[i]
			//We only increment k if we find a unique element. This helps us maintain a spot on where to put the next unique value
			k++
		}
		i++
	}
	return k
}

File: leetcode/RemoveDuplicates.java
package leetcode;

import java.util.TreeSet;

public class RemoveDuplicates {
    public int removeDuplicates(int[] nums) {
        //Two pointers
        int l = 1;
        for(int r = 1; r < nums.length; r++) {
            if(nums[r] != nums[r - 1]) {
                nums[l++] = nums[r];
            }
        }
        return l;
    }

    public int removeDuplicatesSet(int[] nums) {
        TreeSet<Integer> unique = new TreeSet<>();
        for(int num : nums) {
            unique.add(num);
        }

        int i = 0;
        for(int num : unique) {
            nums[i++] = num;
        }
        return i;
    }
}
